rclcpp/test/topic_statistics/test_subscription_topic_statistics.cpp

```diff
@@ -97,21 +99,49 @@ class EmptyPublisher : public rclcpp::Node
 
   virtual ~EmptyPublisher() = default;
 
-  size_t get_number_published()
+private:
+  void publish_message()
+  {
+    auto msg = Empty{};
+    publisher_->publish(msg);
+  }
+
+  rclcpp::Publisher<Empty>::SharedPtr publisher_;
+  rclcpp::TimerBase::SharedPtr publish_timer_;
+};
+
+/**
+ * MessageWithHeader publisher node: used to publish MessageWithHeader with `header` value set
+ */
+class DummyPublisher : public rclcpp::Node
+{
+public:
+  DummyPublisher(
+    const std::string & name, const std::string & topic,
+    const std::chrono::milliseconds & publish_period = std::chrono::milliseconds{100})
+  : Node(name)
   {
-    return number_published_.load();
+    publisher_ = create_publisher<MessageWithHeader>(topic, 10);
+    publish_timer_ = this->create_wall_timer(
+      publish_period, [this]() {
+        this->publish_message();
+      });
   }
 
+  virtual ~DummyPublisher() = default;
+
 private:
   void publish_message()
   {
-    ++number_published_;
-    auto msg = Empty{};
+    auto msg = MessageWithHeader{};
+    auto now = this->now();
+    auto nanos = now.nanoseconds() - 1000 * 1000;
+    msg.header.stamp.sec = static_cast<std::int32_t>(RCL_NS_TO_S(nanos));
+    msg.header.stamp.nanosec = static_cast<std::uint32_t>(nanos % (1000 * 1000 * 1000));
```

Comment:
I think we can avoid the manual conversion and use the assignment operator directly:

```c++
msg.header.stamp = this->now();
```